exonum/src/events/noise/mod.rs

```diff
@@ -0,0 +1,114 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use byteorder::{ByteOrder, LittleEndian};
+use futures::future::{done, Future};
+use tokio_core::net::TcpStream;
+use tokio_io::{AsyncRead, codec::Framed, io::{read_exact, write_all}};
+
+use std::io;
+
+use crypto::{PublicKey, SecretKey};
+use events::codec::MessagesCodec;
+use events::noise::wrapper::{NoiseWrapper, HANDSHAKE_HEADER_LENGTH};
+
+pub mod wrapper;
+
+type HandshakeResult = Box<Future<Item = Framed<TcpStream, MessagesCodec>, Error = io::Error>>;
+
+#[derive(Debug, Clone)]
+/// Params needed to establish secured connection using Noise Protocol.
+pub struct HandshakeParams {
+    pub public_key: PublicKey,
+    pub secret_key: SecretKey,
+    pub max_message_len: u32,
+}
+
+#[derive(Debug)]
+pub struct NoiseHandshake {}
+
+impl NoiseHandshake {
+    pub fn listen(params: &HandshakeParams, stream: TcpStream) -> HandshakeResult {
+        listen_handshake(stream, params)
+    }
+
+    pub fn send(params: &HandshakeParams, stream: TcpStream) -> HandshakeResult {
+        send_handshake(stream, params)
+    }
+}
+
+fn listen_handshake(stream: TcpStream, params: &HandshakeParams) -> HandshakeResult {
+    let max_message_len = params.max_message_len;
+    let mut noise = NoiseWrapper::responder(params);
+    let framed = read(stream).and_then(move |(stream, msg)| {
+        let _buf = noise.read_handshake_msg(&msg);
+        write_handshake_msg(&mut noise)
+            .and_then(|(len, buf)| write(stream, &buf, len))
+            .and_then(|(stream, _msg)| read(stream))
+            .and_then(move |(stream, msg)| {
+                let _buf = noise.read_handshake_msg(&msg);
+                let noise = noise.into_transport_mode()?;
+                let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
+                Ok(framed)
+            })
+    });
+
+    Box::new(framed)
+}
+
+fn send_handshake(stream: TcpStream, params: &HandshakeParams) -> HandshakeResult {
+    let max_message_len = params.max_message_len;
+    let mut noise = NoiseWrapper::initiator(params);
+    let framed = write_handshake_msg(&mut noise)
+        .and_then(|(len, buf)| write(stream, &buf, len))
+        .and_then(|(stream, _msg)| read(stream))
+        .and_then(move |(stream, msg)| {
+            let _buf = noise.read_handshake_msg(&msg);
+            write_handshake_msg(&mut noise)
+                .and_then(|(len, buf)| write(stream, &buf, len))
+                .and_then(move |(stream, _msg)| {
+                    let noise = noise.into_transport_mode()?;
+                    let framed = stream.framed(MessagesCodec::new(max_message_len, noise));
+                    Ok(framed)
+                })
+        });
+
+    Box::new(framed)
+}
+
+fn read(sock: TcpStream) -> Box<Future<Item = (TcpStream, Vec<u8>), Error = io::Error>> {
```

Comment:
Maybe, Rust v1.26 with its `impl Trait` could make this code more natural?

Reply:
I'll change this signature after we move to 1.26.